rs/ledger_suite/icrc1/tokens_u256/src/encode_decode.rs

```diff
@@ -0,0 +1,55 @@
+use ethnum::u256;
+use minicbor::data::Tag;
+use minicbor::decode::{Decoder, Error};
+use minicbor::encode::{Encoder, Write};
+
+const U32_MAX: u256 = u256::new(u32::MAX as u128);
+const U64_MAX: u256 = u256::new(u64::MAX as u128);
+
+pub fn decode<Ctx>(d: &mut Decoder<'_>, _ctx: &mut Ctx) -> Result<u256, Error> {
+    let pos = d.position();
+    match d.u64() {
+        Ok(n) => return Ok(u256::from(n)),
+        Err(e) if e.is_type_mismatch() => {
+            d.set_position(pos);
+        }
+        Err(e) => return Err(e),
+    }
+
+    let tag: Tag = d.tag()?;
+    if tag != Tag::PosBignum {
+        return Err(Error::message(
+            "failed to parse u256: expected a PosBignum tag",
+        ));
+    }
+    let bytes = d.bytes()?;
+    if bytes.len() > 32 {
+        return Err(Error::message(format!(
+            "failed to parse u256: expected at most 32 bytes, got: {}",
+            bytes.len()
+        )));
+    }
+    let mut be_bytes = [0u8; 32];
+    be_bytes[32 - bytes.len()..32].copy_from_slice(bytes);
+    Ok(u256::from_be_bytes(be_bytes))
+}
+
+pub fn encode<Ctx, W: Write>(
+    v: &u256,
+    e: &mut Encoder<W>,
+    _ctx: &mut Ctx,
+) -> Result<(), minicbor::encode::Error<W::Error>> {
+    if v <= &U32_MAX {
+        e.u32(v.as_u32())?;
+    } else if v <= &U64_MAX {
+        e.u64(v.as_u64())?;
+    } else {
+        let be_bytes = v.to_be_bytes();
+        let non_zero_pos = be_bytes
+            .iter()
+            .position(|x| *x != 0)
+            .unwrap_or(be_bytes.len());
+        e.tag(Tag::PosBignum)?.bytes(&be_bytes[non_zero_pos..])?;
+    }
+    Ok(())
+}
```

Comment:
is there a way to avoid a custom implementation here, e.g., by converting to a big number type and use its implementation for deserialization/serialization?

Reply:
This file is an exact copy of https://github.com/dfinity/ic/blob/bef38d0770b97e74e40d0530ed95b6413df2f8b2/rs/ethereum/cketh/minter/src/cbor/u256.rs There is a ticket to make it available as a library: https://dfinity.atlassian.net/browse/XC-48 Once that is done, I will use that library and delete this file. I didn't investigate the possibility to use big number serialization...

Reply:
if the file is already checked in the repository, can it be maybe reshared at least within dfinity/ic?

Reply:
@gregorydemay would there be a way to use the file in the ICRC1 ledger code?

Reply:
@maciejdfinity : sure, make it a library 😄 . I can check if I have some time tomorrow for this.